### ISSUE TITLE
Error in logs: error invoke actor method: failed to cast response

### ISSUE THREAD

**Issue by IvanILanekassen** [dapr/dapr#6339]
## In what area(s)?
/area runtime

## What version of Dapr?
1.10.5

## Expected Behavior

No errors of this kind should happen

## Actual Behavior

We see a lot of Dapr.DaprApiException excptions in our logs after upgrading from 1.9.5 to 1.10.5. The exceptions have the following error message: "error invoke actor method: failed to cast response". This exact error message returns only 1 hit on google, namely the [dapr source code that returns it](https://github.com/dapr/dapr/blob/770d4e51604f1264d8bb25cedf16ea9f77539394/pkg/http/api.go#L1863). We have investigated some of the occurences, and see that the issue occurs when there is some kind of error in the actor. The error results in an exception thrown from the actor. I tried to browse the docs to see what is expected/acceptable return types from the actor, but could not find anything. Right now, our actor methods return Task (but might throw an exception). We are on v1.10.0 of the .NET SDK (Dapr.Actors.AspNetCore). So, should we not throw exceptions? How should the actors behave in the case of an error?

## Steps to Reproduce the Problem

Unsure if it is possible, but create an actor that throws an exception, and invoke this actor using the .NET SDK. This might trigger the error. 

## Release Note

RELEASE NOTE: **ADD** New documentation on error handling in actors.

**Comment by nissbran** [comment]
This seems to be related to the changes made in the #5645 pull request. The policy runner sets the response to zero at https://github.com/dapr/dapr/blob/1a79c6f2de62a20290f5a399199a3fb5e31dcf2d/pkg/resiliency/policy.go#L195

This only affects the .NET SDK implementation as I understand it. Because it sends errors in the response as 200 and with a header.

**Comment by akkie** [comment]
We faced the same issue and for us the root cause was that we had two actor implementations with the same name in two different services. It seems that when calling the remote actor, that Dapr calls the actor from a different service and then the above error occurs.

I know it's an error on our side. But maybe the info is usefull for others as well.

**Comment by artursouza** [comment]
Does this issue also happen with 1.10.9?

**Comment by onionhammer** [comment]
Yes

**Comment by onionhammer** [comment]
Just got it in production today, wonder what the error was...  guess I'll never know because failed to cast response.

**Comment by tmacam** [comment]
I can repro it with a brand new AKS cluster running Dapr 1.11.2 using the current Bicep for our longhaul E2E tests from https://github.com/dapr/test-infra#on-azure-kubernetes-service-aks

```
│Increasing star_verypositive.                                                                                                                                                                                                                                                                                                                    
 Increasing star_verypositive failed with Dapr.DaprApiException: error invoke actor method: failed to cast response                                                                                                                                                                                                                               
    at Dapr.Actors.DaprHttpInteractor.SendAsyncHandleUnsuccessfulResponse(Func`1 requestFunc, String relativeUri, CancellationToken cancellationToken)                                                                                                                                                                                            
    at Dapr.Actors.DaprHttpInteractor.SendAsync(Func`1 requestFunc, String relativeUri, CancellationToken cancellationToken)                                                                                                                                                                                                                      
    at Dapr.Actors.DaprHttpInteractor.InvokeActorMethodWithRemotingAsync(ActorMessageSerializersManager serializersManager, IActorRequestMessage remotingRequestRequestMessage, CancellationToken cancellationToken)                                                                                                                              
    at Dapr.Actors.Communication.Client.ActorRemotingClient.InvokeAsync(IActorRequestMessage remotingRequestMessage, CancellationToken cancellationToken)                                                                                                                                                                                         
    at Dapr.Actors.Client.ActorProxy.InvokeMethodAsync(Int32 interfaceId, Int32 methodId, String methodName, IActorRequestMessageBody requestMsgBodyValue, CancellationToken cancellationToken)                                                                                                                                                   
    at Dapr.Tests.HashTagApp.Controllers.HashTagController.PostMessageBinding(SocialMediaMessage message) in /app/hashtag-counter/Controllers/HashTagController.cs:line 74                                                                                                                                                                        
 fail: Microsoft.AspNetCore.Server.Kestrel[13]                                                                                                                                                                                                                                                                                                    
       Connection id "0HMTGPO25605S", Request id "0HMTGPO25605S:00000015": An unhandled exception was thrown by the application.                                                                                                                                                                                                                  
       Dapr.DaprApiException: error invoke actor method: failed to cast response                                                                                                                                                                                                                                                                  
          at Dapr.Tests.HashTagApp.Controllers.HashTagController.PostMessageBinding(SocialMediaMessage message) in /app/hashtag-counter/Controllers/HashTagController.cs:line 74                                                                                                                                                                  
          at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)                                                                                                  
          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)                                                                                                                                                         
          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)                                                                                                             
          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)                                                                                                                                                                                                                         
          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)                                                                                                                                                                                                 
          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)                                                                                                                  
          at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)                                                                                                                               
          at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)                                                                                                                                                                                          
          at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)                                                                                                                                                                                          
          at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)                                                                                                                                                                                                 
          at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)                                                                                                                                                                                                                                               
          at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)      
```

**Comment by fabistb** [comment]
A short side note since I have seen this with dapr 1.10.7 and the servicebus pub/sub component from time to time.
In the logs the here mentioned exception appeared frequently, however one of the retries normally had an exception which actually contained more useful information.

**Comment by fabistb** [comment]
> > A short side note since I have seen this with dapr 1.10.7 and the servicebus pub/sub component from time to time.
> > In the logs the here mentioned exception appeared frequently, however one of the retries normally had an exception which actually contained more useful information.
> 
> @fabistb You saw the same error in pub/sub component? Was it during publish or subscribe?
> 
> Could you paste the exception that you saw by chance?

@mukundansundar sorry should have been more precise.
I have seen this in the combination pub/sub and actor.

So message is picked, actor instance is created and processing ist started.
An exception is thrown and because of that the retries are executed.

**Comment by artursouza** [comment]
Is this exception shown consistently or is it intermittent?

**Comment by artursouza** [comment]
> 

@halspang Can you see if this is related to .Net SDK?

**Comment by artursouza** [comment]
> This seems to be related to the changes made in the #5645 pull request. The policy runner sets the response to zero at
> 
> https://github.com/dapr/dapr/blob/1a79c6f2de62a20290f5a399199a3fb5e31dcf2d/pkg/resiliency/policy.go#L195
> 
> This only affects the .NET SDK implementation as I understand it. Because it sends errors in the response as 200 and with a header.
> 
> The reason it only affects some method calls is that this only occurs when a remoting actor call is made. So if the actor runs on the same host as the caller it works as intended.

@halspang Can you see if this is related to .Net SDK? It seems to be a problem with remoting actors.

**Comment by mukundansundar** [comment]
> > This seems to be related to the changes made in the #5645 pull request. The policy runner sets the response to zero at
> > https://github.com/dapr/dapr/blob/1a79c6f2de62a20290f5a399199a3fb5e31dcf2d/pkg/resiliency/policy.go#L195
> > 
> > This only affects the .NET SDK implementation as I understand it. Because it sends errors in the response as 200 and with a header.
> > The reason it only affects some method calls is that this only occurs when a remoting actor call is made. So if the actor runs on the same host as the caller it works as intended.
> 
> @halspang Can you see if this is related to .Net SDK? It seems to be a problem with remoting actors.

This specifically happens when the response is set to nil/disposed as seen in the code above `rRes=zero` , the following code block then throws the error 
https://github.com/dapr/dapr/blob/b76a49f08c09664ce9c7068373d4ec810e3140a7/pkg/http/api.go#L1820-L1835

**Comment by ItalyPaleAle** [comment]
@tmacam you said you had code that repros this 100% of the time? Would you be able to share that?

**Comment by tmacam** [comment]
> @tmacam you said you had code that repros this 100% of the time? Would you be able to share that?

Yes. It is the code under https://github.com/dapr/test-infra. If you bring a cluster as described in https://github.com/dapr/test-infra/blob/master/README.md#on-azure-kubernetes-service-aks the resulting hashtag-counter-app has this behaviour. I am trying to check if I can repro this locally

**Comment by artursouza** [comment]
I wrote a simple E2E test scenario that reproduces the issue. I validated that the error is not present at 1.9.6. I am validating other versions now.

**Comment by artursouza** [comment]
I confirmed the E2E test reproduces the issue also in 1.10.9.

**Comment by artursouza** [comment]
I validated the error also requires a resiliency policy when invoking actors.

**Comment by onionhammer** [comment]
It definitely happens frequently on ACA, and [ACA does not support resiliency](https://github.com/microsoft/azure-container-apps/issues/585)

**Comment by ItalyPaleAle** [comment]
There are built-in resiliency policies for actor invocation

**Comment by artursouza** [comment]
I disabled resiliency and it worked. Put it back and it failed. Anyway, the PR with the fix is almost ready.

**Comment by Will-Dav** [comment]
@artursouza , apologies, I'm new to Dapr. How do I disable resiliency?

**Comment by KieranMcCormick** [comment]
@artursouza just following up, any chance you can explain what you meant by disabling resiliency?

We cant jump to the latest version so a work around would be much appreciated!

**Comment by artursouza** [comment]
> @artursouza , apologies, I'm new to Dapr. How do I disable resiliency?

You can comment out the policies (in the YAML file) that apply to actors and that should take care of it.

**Comment by Will-Dav** [comment]
> You can comment out the policies (in the YAML file) that apply to actors and that should take care of it.

Thanks Artur, I appreciate you getting back to me.
I'm using the default config file, which doesn't appear to have anything related to resiiency or actors.

apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: daprConfig
spec:
  tracing:
    samplingRate: "1"
    zipkin:
      endpointAddress: http://localhost:9411/api/v2/spans
  features:
    - name: proxy.grpc
      enabled: true

**Comment by artursouza** [comment]
Fixed in 1.12 and 1.11. Pending fix in 1.10

**Comment by onionhammer** [comment]
@artursouza is there a related PR?
